### ISSUE TITLE
vim.ui_attach: detect if a message was sent because of a redraw (with `ext_messages`)

### ISSUE THREAD

**Issue by folke** [neovim/neovim#20416]
### Feature description

Whenever a `redraw` happens anywhere, the non-cleared messages will be resent.

This is problematic, if the code that handles messages, also does a redraw. Redraws in my code only happen when any queued messages have not been rendered yet and we're in some sort of blocking call.

This would effectively cause some sort of loop.

To circumvent this, I currently [wrap any redraw function](https://github.com/folke/noice.nvim/blob/dee23c21c3ff52119d8452b77814e6be75dd23bb/lua/noice/hacks.lua#L55) and when we're inside a redraw cycle, we simply ignore incoming messages.

Not sure what the best way would be to solve this for `ext_messages`.

**Comment by bfredl** [comment]
ext_messages should not need to clear and then resend messages due to redraw in the first place. this is mostly a consequence of naively following the old vim control flow (with constant clear/redraw panic).

**Comment by folke** [comment]
Yeah, that would be even better for sure. It's confusing for API consumers.

**Comment by folke** [comment]
@luukvbaal this is probably fixed now with your recent changes? Since callbacks can only use fast events?

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/31224

**Comment by folke** [comment]
Right, so if a redraw happens somewhere, the re-emit still happens and there's no way to detect this is a resend right?
Some kind of flag in the `msg_show` or msg id to know it has already been processed would be great.

Currently I just discard msg_show events if they're the same as the last one, but this also means if you echo the same message twice, Noice will only show it once.
